tests/test_pluvo.py

```diff
@@ -17,112 +17,99 @@ def results(self, *args, **kwargs):
+        call('GET', 'endpoint', params={'limit': 2, 'offset': 2}),
     ])
 
 
-def test_pluvo_generator_limit(mocker):
-    pages = [
-        {'count': 4, 'data': [1, 2]},
-        {'count': 4, 'data': [3, 4]}
-    ]
-
-    _request_mock = mocker.MagicMock(side_effect=Multiple(pages).results)
-    pluvo_mock = mocker.MagicMock(page_size=2, _request=_request_mock)
+def test_pluvo_resultset_get_page_key_offset(mocker):
+    pluvo_mock = mocker.MagicMock(page_size=2)
+    p = PluvoResultSet(pluvo_mock, 'endpoint')
 
-    retval = PluvoGenerator(pluvo_mock, 'endpoint', params={'limit': 3})
-
-    assert len(retval) == 3
-    assert list(retval) == [1, 2, 3]
-    _request_mock.assert_has_calls([
-        call('GET', 'endpoint', params={'limit': 2, 'offset': 0}),
-        call('GET', 'endpoint', params={'limit': 1, 'offset': 2})
-    ])
+    assert p._get_page_key_offset(0) == (0, 0)
+    assert p._get_page_key_offset(1) == (0, 1)
+    assert p._get_page_key_offset(7) == (3, 1)
 
 
-def test_pluvo_generator_offset(mocker):
+def test_pluvo_resultset_getitem(mocker):
     pages = [
-        {'count': 4, 'data': [3, 4]}
+        [0, 1],
+        [2, 3],
+        [4, 5],
+        [6, 7],
     ]
 
-    _request_mock = mocker.MagicMock(side_effect=Multiple(pages).results)
-    pluvo_mock = mocker.MagicMock(page_size=2, _request=_request_mock)
-
-    retval = PluvoGenerator(pluvo_mock, 'endpoint', params={'offset': 2})
-
-    assert len(retval) == 2
-    assert list(retval) == [3, 4]
-    _request_mock.assert_has_calls([
-        call('GET', 'endpoint', params={'limit': 2, 'offset': 2})
-    ])
-
-
-def test_pluvo_generator_limit_and_offset(mocker):
+    get_page_mock = mocker.MagicMock(side_effect=lambda x: pages[x])
+    pluvo_mock = mocker.MagicMock(page_size=2)
+    p = PluvoResultSet(pluvo_mock, 'endpoint')
+    p._count = 8
+    with patch.object(p, '_get_page', get_page_mock):
+        assert p[:] == [0, 1, 2, 3, 4, 5, 6, 7]
+        assert p[-1] == 7
+        assert p[:1] == [0]
+        assert p[-3:] == [5, 6, 7]
+        assert p[8:1] == []
+        assert p[0:1] == [0]
+        assert p[3:7] == [3, 4, 5, 6]
+        with pytest.raises(IndexError):
+            p[8]
+
+
+def test_pluvo_resultset_len(mocker):
     pages = [
-        {'count': 1, 'data': [3]}
+        {'count': 4, 'data': [0, 1]},
+        {'count': 4, 'data': [2, 3]}
     ]
+    i = iter(pages)
 
-    _request_mock = mocker.MagicMock(side_effect=Multiple(pages).results)
-    pluvo_mock = mocker.MagicMock(page_size=2, _request=_request_mock)
+    request_mocker = mocker.MagicMock(
+        side_effect=lambda *args, **kwargs: next(i))
+    pluvo_mock = mocker.MagicMock(page_size=2, _request=request_mocker)
+    p = PluvoResultSet(pluvo_mock, 'endpoint')
 
-    retval = PluvoGenerator(pluvo_mock, 'endpoint',
-                            params={'limit': 1, 'offset': 2})
+    assert len(p) == 4
+    assert len(p) == 4
+    assert pluvo_mock._request.call_count == 1
```

Comment:
It's probably a good idea to add a comment here, because otherwise this test will still pass after someone deletes the second `len` call. Hell, it might even be a good idea to do something like `assert p.__len__.call_count == 2`, just to prevent someone from removing that second call.